packages/python/src/components/DataclassDeclaration.tsx

```diff
@@ -0,0 +1,212 @@
+    );
+  }
+
+  // Enforce at most one KW_ONLY sentinel as a symbol
+  let kwOnlyCount = 0;
+  for (const sym of owner.instanceMembers as Iterable<any>) {
+    if (sym.originalName === "_") kwOnlyCount++;
+  }
+  if (kwOnlyCount > 1) {
+    throw new Error("Only one KW_ONLY sentinel is allowed per dataclass body");
+  }
+}
+
+/**
+ * Allowed keyword arguments for the Python `@dataclass(...)` decorator.
+ * Single source of truth: runtime keys and compile-time type are derived here.
+ */
+export const dataclassDecoratorKeys = [
+  "init",
+  "repr",
+  "eq",
+  "order",
+  "unsafeHash",
+  "frozen",
+  "matchArgs",
+  "kwOnly",
+  "slots",
+  "weakrefSlot",
+] as const;
+export type DataclassDecoratorKey = (typeof dataclassDecoratorKeys)[number];
+export type DataclassDecoratorKwargs = Partial<
+  Record<DataclassDecoratorKey, boolean>
+>;
+
+export interface DataclassDeclarationProps
+  extends ClassDeclarationProps,
+    DataclassDecoratorKwargs {}
+
+/**
+ * Renders a Python dataclass. Uses ClassDeclaration component internally.
+ *
+ * Example:
+ * ```tsx
+ * <py.DataclassDeclaration name="User" kwOnly>
+ *   <py.VariableDeclaration instanceVariable omitNone name="id" type="int" />
+ *   <py.VariableDeclaration instanceVariable name={namekey("_", { ignoreNamePolicy: true })} type={dataclassesModule["."].KW_ONLY} omitNone />
+ *   <py.VariableDeclaration
+ *     instanceVariable
+ *     name="name"
+ *     type="str"
+ *     initializer={"Anonymous"}
+ *   />
+ * </py.DataclassDeclaration>
+ * ```
+ * Will render as:
+ * ```py
+ * from dataclasses import dataclass
+ * from dataclasses import KW_ONLY
+ *
+ * @dataclass(kw_only=True)
+ * class User:
+ *     id: int
+ *     _: KW_ONLY
+ *     name: str = "Anonymous"
+ * ```
+ */
+export function DataclassDeclaration(props: DataclassDeclarationProps) {
+  const decoratorKeys: (keyof DataclassDecoratorKwargs)[] = [
+    ...dataclassDecoratorKeys,
+  ];
+  const validKeySet = new Set<string>(decoratorKeys as unknown as string[]);
+  // Collect flags from props in the order they appear (preserves emission order)
+  const orderedKwargs: Array<[keyof DataclassDecoratorKwargs, any]> = [];
+  for (const key of Object.keys(props)) {
+    // Only include known flags; skip undefined values
+    if (validKeySet.has(key)) {
+      const value = (props as any)[key];
+      if (value !== undefined)
+        orderedKwargs.push([key as keyof DataclassDecoratorKwargs, value]);
+    }
+  }
+  // Materialize ordered entries into an object for validation/rendering
+  const kwargs = orderedKwargs.reduce((acc, [k, v]) => {
+    (acc as any)[k] = v;
+    return acc;
+  }, {} as DataclassDecoratorKwargs);
+  const hasDecoratorArgs = orderedKwargs.length > 0;
+  const toSnakeCase = (s: string): string =>
+    s
+      .replace(/([a-z\d])([A-Z])/g, "$1_$2")
+      .replace(/([A-Z]+)([A-Z][a-z])/g, "$1_$2")
+      .toLowerCase();
+  const decoratorEntries = orderedKwargs.map(([k, v]) => {
```

Comment:
Recase when you assemble orderedKwargs, it seems like we don't need the camelCased version for anything so might as well do it in one pass.